include/swift/IDE/Utils.h

```diff
@@ -128,71 +128,201 @@ enum class CursorInfoKind {
   StmtStart,
 };
 
+/// Base class of more specialized \c ResolvedCursorInfos that also represents
+/// and \c Invalid cursor info.
+/// Subclasses of \c ResolvedCursorInfo cannot add new stored properies because
+/// \c ResolvedCursorInfo is being passed around as its base class and thus any
+/// properties in subclasses would get lost.
 struct ResolvedCursorInfo {
+protected:
   CursorInfoKind Kind = CursorInfoKind::Invalid;
   SourceFile *SF = nullptr;
   SourceLoc Loc;
-  ValueDecl *ValueD = nullptr;
-  TypeDecl *CtorTyRef = nullptr;
-  ExtensionDecl *ExtTyRef = nullptr;
-  /// Declarations that were shadowed by \c ValueD using a shorthand syntax that
-  /// names both the newly declared variable and the referenced variable by the
-  /// same identifier in the source text. This includes shorthand closure
-  /// captures (`[foo]`) and shorthand if captures
-  /// (`if let foo {`).
-  /// Decls that are shadowed using shorthand syntax should be reported as
-  /// additional cursor info results.
-  SmallVector<ValueDecl *, 2> ShorthandShadowedDecls;
-  ModuleEntity Mod;
-  bool IsRef = true;
-  bool IsKeywordArgument = false;
-  Type Ty;
-  Type ContainerType;
-  Stmt *TrailingStmt = nullptr;
-  Expr *TrailingExpr = nullptr;
-  /// It this is a ref, whether it is "dynamic". See \c ide::isDynamicRef.
-  bool IsDynamic = false;
-  /// If this is a dynamic ref, the types of the base (multiple in the case of
-  /// protocol composition).
-  SmallVector<NominalTypeDecl *, 1> ReceiverTypes;
 
+  // Technically, these structs could form a union (because only one of them is
+  // active at a time). But I had issues with C++ complaining about copy
+  // constructors and gave up. At the moment it's only wasting 3 words for non
+  // ValueRef data.
+  struct {
+    ValueDecl *ValueD = nullptr;
+    TypeDecl *CtorTyRef = nullptr;
+    ExtensionDecl *ExtTyRef = nullptr;
+    bool IsRef = true;
+    Type Ty;
+    Type ContainerType;
+    bool IsKeywordArgument = false;
+    /// It this is a ref, whether it is "dynamic". See \c ide::isDynamicRef.
+    bool IsDynamic = false;
+    /// If this is a dynamic ref, the types of the base (multiple in the case of
+    /// protocol composition).
+    SmallVector<NominalTypeDecl *, 1> ReceiverTypes;
```

Comment:
While you're here you may as well remove the default from both of these and let `SmallVector` use whatever fits instead.

Reply:
I forgot that `SmallVector` now has a default for its size.